### DISCUSSION TITLE
SD-XL ControlNet released as part of Diffusers

### DISCUSSION THREAD

**Discussion by Alchete** [vladmandic/sdnext#1982] [General]
It looks like the Canny model has been released.  Perhaps others to follow.
https://huggingface.co/diffusers/controlnet-sdxl-1.0

The discussion page details how someone got it to work with Comfy.  Perhaps that recipe may help here?

**Comment by vladmandic** [comment]
its planned, but non trivial as adding a model is one thing, but it also needs a proper front end to be usable.

**Reply by wandrzej** [reply]
Can you elaborate, what is needed on front end to make it work with A1111? Are the inputs somehow different compared to SD1X? I thought it was more a matter of architecture

**Reply by vladmandic** [reply]
Existing controlnet extension does not work with diffusers at all, so entire front end of that extension should be redone natively instead of relying on extension.

**Comment by radry** [comment]
So since the controlnet extension doesn't work with diffusers, is it possible to run sdxl without diffusers?

**Reply by midcoastal** [reply]
Yes, use `backend.ORIGINAL`

**Comment by gohan2091** [comment]
Is ControlNet XL suppose to be working so far in Vlads A1111? I can't get it to work. I have the XL models Canny and Depth installed and showing up but the generation ignores the control nets.

**Comment by radry** [comment]
Aparently the controlnet extension DOES work with diffusers and SDXL, so why isn't it working in SD.Next?
https://github.com/Mikubill/sd-webui-controlnet/discussions/2039

**Reply by vladmandic** [reply]
Search. It's been discussed many times.

**Reply by radry** [reply]
Yeah I found this issue by you and the fix of the extension. Still when updating and enabling the extension in SD.next, it gets automatically disabled.  
https://github.com/Mikubill/sd-webui-controlnet/issues/2040
https://github.com/Mikubill/sd-webui-controlnet/pull/2041

**Reply by gohan2091** [reply]
Yes, My controlnet is always disabled. When I retick it and restart server, it remains disabled. Only when I switch back to SD 1.5 settings does Controlnet become enabled.

**Reply by vladmandic** [reply]
Yes, exactly.

**Comment by gohan2091** [comment]
> Yes, exactly.

? Can I not use ControlNet with SD XL in SD.Net?

**Reply by vladmandic** [reply]
not yet. that has been stated many many times. its work in progress.

**Reply by gohan2091** [reply]
Really sorry, it must be frustrating saying the same thing over and over. I don't read this page so often and things are sometimes difficult to locate. Thanks for clarifying. It's my understanding that the default A1111 does support ControlNet XL. Please correct me if I am wrong.

**Reply by vladmandic** [reply]
it does since few days ago, so its not that much of a difference.

**Reply by midcoastal** [reply]
@vladmandic - Hate to be that way, but now it's been a few months.  Anything worth looking at?

**Reply by vladmandic** [reply]
There is a working prototype that is being evaluated. Once it's in suitable state, it will be merged.

Amount of work to make this happen is staggering.
You'll see once code is published.

**Reply by midcoastal** [reply]
Oh, I'm sure.

Off hand, why was it so much different than the A1111 implimentation (IE: why couldn't the code from A be used, or at least used as boilerplate, here?)

**Reply by vladmandic** [reply]
sure it can be used as boiler-plate, but sd-webui-controlnet extension is by far the biggest piece of code other than core repos. its 1.2k files (769 python source files).

**Reply by midcoastal** [reply]
.......... o.0

**Reply by vladmandic** [reply]
note that diffusers do offer "simpler" way of using controlnet pipeline, but that has no value unless you have preprocessed images as inputs - and each preprocessor is a different model that requires different handling.

**Reply by midcoastal** [reply]
@vladmandic - Looking forward to seeing this launched.  It just occurred to me that `IP Adapter` is reliant on `ControlNet` (requires the `load_ip_adapter` method that, best I can see, is only provided in the `ControlNet` Package), and therefore does not work in `backend.DIFFUSERS`.  Womp womp...  Unless I am missing something...

**Reply by vladmandic** [reply]
its not, ip adapters are being added to standard pipelines in next version of diffusers and its already in sdnext dev branch, you just need to install diffusers from their main branch.

**Reply by blacklig** [reply]
so how do you even implement those difufsers CN pipelines if they are that much simpler? And I know they are.. I was wondering how would you add those "start from XY step" etc if this even is not implemented in diffusers I think, or is it somehow?

**Reply by vladmandic** [reply]
> so how do you even implement those difufsers CN pipelines if they are that much simpler? And I know they are.. I was wondering how would you add those "start from XY step" etc if this even is not implemented in diffusers I think, or is it somehow?

I'm not going into full how-to here, it would be a veeeey long writeup. You'll see the code once it's merged to dev.

**Reply by blacklig** [reply]
yup, no problem.. My point was if you like "re implemented" solution diffusers use in more advanced way or just somwhow twisted those pipelines? As I was playing with CN pipelines and they really seems like less advanced than what you normaly offer in SD Next for 1.5 CNs.

**Reply by vladmandic** [reply]
diffusers pipelines are not less advanced, they are just different. there is a lot you can via things like `callback_on_step_end_tensor_inputs`

**Reply by blacklig** [reply]
ah, ok. So you are basically not using diffusers pipe lines but you are calling same stuff like diffusers pipe lines are calling under the hood? Only you call it in the way you need it to CN properly work and get all those settings we know from 1.5 implementation in Next?

**Reply by vladmandic** [reply]
i am using pipelines, but i via pipeline callbacks, you can modify pipeline internals on-the-fly.

**Comment by fcabanski** [comment]
Thank goodness it's coming soon. A1111 is OK, but SDNext is much better. Once it has Controlnet for SDNext, I can get back to using it.

Vlad is a wonder man!
